packages/synapse-react-client/src/utils/hooks/useDetectSSOCode.ts

```diff
@@ -59,11 +59,33 @@ export default function useDetectSSOCode(
   const { searchParams } = fullUrl
   const code = searchParams?.get('code')
   const provider = searchParams?.get('provider')
-  // state is used during OAuth based Synapse account creation (it's the username)
-  const encodedState = searchParams?.get('state')
-  const state = encodedState
-    ? (JSON.parse(decodeURIComponent(encodedState)) as OAuth2State)
-    : null
+
+  // If the URL contains a client_id and redirect_uri, then we are acting as an identity provider for an external OAuth client
+  const isHandlingSynapseOAuthSignIn = Boolean(
+    searchParams?.get('client_id') && searchParams?.get('redirect_uri'),
+  )
+
+  // If the Synapse user signed in with an external IdP, we may have passed data in the 'state' param
+  // Parse it (if appropriate)
+  const state: OAuth2State | null = useMemo(() => {
+    // If we are acting as an OIDC identity provider, then we should not parse the state param -- it was sent to us, and we should return it untouched
+    if (!isHandlingSynapseOAuthSignIn) {
+      const encodedState = searchParams?.get('state')
+      try {
+        return encodedState
+          ? (JSON.parse(decodeURIComponent(encodedState)) as OAuth2State)
+          : null
+      } catch (e) {
+        console.error(
+          'Error parsing state param:\n',
+          e,
+          '\nEncoded value:\n',
+          encodedState,
+        )
+      }
+    }
+    return null
+  }, [isHandlingSynapseOAuthSignIn, searchParams])
```

Comment:
WDYT of this logic? Try to guess if we're in the middle of an OAuth sign-in attempt where we are the IdP, and if so, do not parse the state.

Reply:
I wonder if we can separate out the concerns for this app.  I think this is fine for now, but what if in the future we delegated establishing a Synapse session to OneSage?  The redirectURL that we would send OneSage would be the current URL (including the state that the client sent to the Synapse IdP).  Would that simplify this solution (if we took out the login component)?  How bad do you think the UX would be with this additional redirect (if a session is not discovered)?  I like the idea that this sign-in app would not need to worry about other login complexities (like the 2FA reset workflow!).  WDYT?

Reply:
I agree, I think separating concerns per-app would be ideal. It's a lot of redirects, but I think it's pretty standard for systems that federate identity + consent. We could consider combining both the OAuth signin + OneSage login+account management into an individual SPA to save a few full page loads, but then we'd have to put OneSage on the signin.synapse.org domain.